File: packages/sdk-ts/src/core/pagination.ts
import { APIPromise } from './api-promise'
import type { APIResponse, ResponseMeta } from './http'

/**
 * Cursor pagination.
 *
 * `list()` returns one page — `{ data, nextCursor, hasMore }` — and the same
 * object walks every page when you loop over it:
 *
 *   const page = await invoiceai.invoices.list({ limit: 50 })   // one page
 *   for await (const inv of invoiceai.invoices.list())           // every invoice
 *   const first500 = await invoiceai.invoices.list().toArray({ limit: 500 })
 *
 * Endpoints that return a plain array (no `next_cursor`) produce a single page
 * with `hasMore: false`, so the same code works everywhere.
 */

/** Loads the page after `cursor`. */
export type PageFetcher<Item> = (cursor: string) => PagePromise<Item>

export interface ToArrayOptions {
  /** Stop after this many items in total. */
  limit?: number
}

export class Page<Item> implements AsyncIterable<Item> {
  /** The items on this page. */
  readonly data: Item[]
  /** Pass as `cursor` to get the next page; null on the last page. */
  readonly nextCursor: string | null
  readonly hasMore: boolean
  /** Response metadata of the request that loaded this page. */
  readonly response: ResponseMeta

  constructor(
    body: unknown,
    response: ResponseMeta,
    private readonly fetchNext: PageFetcher<Item>,
  ) {
    const b = (body ?? {}) as { data?: unknown; next_cursor?: unknown }
    this.data = Array.isArray(b.data) ? (b.data as Item[]) : []
    this.nextCursor = typeof b.next_cursor === 'string' && b.next_cursor !== '' ? b.next_cursor : null
    this.hasMore = this.nextCursor !== null
    this.response = response
  }

  hasNextPage(): boolean {
    return this.hasMore
  }

  /** The next page, or null when this is the last one. */
  async getNextPage(): Promise<Page<Item> | null> {
    return this.nextCursor === null ? null : this.fetchNext(this.nextCursor)
  }

  /** This page and every page after it. */
  async *iterPages(): AsyncGenerator<Page<Item>, void, undefined> {
    yield this
    let page = await this.getNextPage()
    while (page) {
      yield page
      page = await page.getNextPage()
    }
  }

  /** Every item on this page and the pages after it. */
  async *[Symbol.asyncIterator](): AsyncGenerator<Item, void, undefined> {
    for await (const page of this.iterPages()) yield* page.data
  }

  /** Collects items across pages, up to `limit` if given. */
  async toArray(options: ToArrayOptions = {}): Promise<Item[]> {
    return collect(this, options.limit)
  }
}

/** A pending page: await it for one page, or `for await` it for every item. */
export class PagePromise<Item> extends APIPromise<Page<Item>> implements AsyncIterable<Item> {
  constructor(responsePromise: Promise<APIResponse>, fetchNext: PageFetcher<Item>) {
    super(responsePromise, (res) => new Page<Item>(res.body, res.meta, fetchNext))
  }

  async *[Symbol.asyncIterator](): AsyncGenerator<Item, void, undefined> {
    const page = await this
    yield* page
  }

  /** Collects items across pages, up to `limit` if given. */
  async toArray(options: ToArrayOptions = {}): Promise<Item[]> {
    return collect(this, options.limit)
  }
}

async function collect<Item>(source: AsyncIterable<Item>, limit: number | undefined): Promise<Item[]> {
  const out: Item[] = []
  if (limit !== undefined && limit <= 0) return out
  for await (const item of source) {
    out.push(item)
    if (limit !== undefined && out.length >= limit) break
  }
  return out
}

File: packages/sdk-python/invoice_ai/pagination.py
"""
Cursor pagination.

``list()`` returns one page (``data``, ``next_cursor``, ``has_more``), and the
same object walks every page when you loop over it:

    page = client.invoices.list(limit=50)                 # one page
    for inv in client.invoices.list():                    # every invoice
        ...
    first_500 = client.invoices.list().to_list(limit=500)

Async:

    page = await client.invoices.list(limit=50)           # one page
    async for inv in client.invoices.list():              # every invoice
        ...
    first_500 = await client.invoices.list().to_list(limit=500)

Endpoints that return a plain array (no ``next_cursor``) produce a single page
with ``has_more == False``, so the same code works everywhere.
"""

from __future__ import annotations

from typing import (
    Any,
    AsyncIterator,
    Awaitable,
    Callable,
    Generator,
    Generic,
    Iterator,
    List,
    Optional,
    Tuple,
    TypeVar,
)

from ._response import ResponseMeta

__all__ = ["AsyncPage", "AsyncPagePromise", "SyncPage"]

T = TypeVar("T")


def _parse_body(body: Any, parse_item: Callable[[Any], T]) -> Tuple[List[T], Optional[str]]:
    b = body if isinstance(body, dict) else {}
    raw = b.get("data")
    data = [parse_item(item) for item in raw] if isinstance(raw, list) else []
    cursor = b.get("next_cursor")
    return data, cursor if isinstance(cursor, str) and cursor != "" else None


class _BasePage(Generic[T]):
    #: The items on this page.
    data: List[T]
    #: Pass as ``cursor`` to get the next page; None on the last page.
    next_cursor: Optional[str]
    has_more: bool
    #: Response metadata of the request that loaded this page.
    response: ResponseMeta

    def __init__(self, body: Any, response: ResponseMeta, parse_item: Callable[[Any], T]) -> None:
        self.data, self.next_cursor = _parse_body(body, parse_item)
        self.has_more = self.next_cursor is not None
        self.response = response

    def has_next_page(self) -> bool:
        return self.has_more

    def __repr__(self) -> str:
        return f"<{type(self).__name__} items={len(self.data)} next_cursor={self.next_cursor!r}>"


class SyncPage(_BasePage[T]):
    """One page. Iterating it yields every item on this page and the pages after it."""

    def __init__(
        self,
        body: Any,
        response: ResponseMeta,
        parse_item: Callable[[Any], T],
        fetch_next: Callable[[str], SyncPage[T]],
    ) -> None:
        super().__init__(body, response, parse_item)
        self._fetch_next = fetch_next

    def get_next_page(self) -> Optional[SyncPage[T]]:
        """The next page, or None when this is the last one."""
        return None if self.next_cursor is None else self._fetch_next(self.next_cursor)

    def iter_pages(self) -> Iterator[SyncPage[T]]:
        """This page and every page after it."""
        page: Optional[SyncPage[T]] = self
        while page is not None:
            yield page
            page = page.get_next_page()

    def __iter__(self) -> Iterator[T]:
        for page in self.iter_pages():
            yield from page.data

    def to_list(self, limit: Optional[int] = None) -> List[T]:
        """Collects items across pages, up to ``limit`` if given."""
        out: List[T] = []
        if limit is not None and limit <= 0:
            return out
        for item in self:
            out.append(item)
            if limit is not None and len(out) >= limit:
                break
        return out


class AsyncPage(_BasePage[T]):
    """One page. ``async for`` over it yields every item on this page and the pages after it."""

    def __init__(
        self,
        body: Any,
        response: ResponseMeta,
        parse_item: Callable[[Any], T],
        fetch_next: Callable[[str], Awaitable[AsyncPage[T]]],
    ) -> None:
        super().__init__(body, response, parse_item)
        self._fetch_next = fetch_next

    async def get_next_page(self) -> Optional[AsyncPage[T]]:
        """The next page, or None when this is the last one."""
        return None if self.next_cursor is None else await self._fetch_next(self.next_cursor)

    async def iter_pages(self) -> AsyncIterator[AsyncPage[T]]:
        """This page and every page after it."""
        page: Optional[AsyncPage[T]] = self
        while page is not None:
            yield page
            page = await page.get_next_page()

    async def __aiter__(self) -> AsyncIterator[T]:
        async for page in self.iter_pages():
            for item in page.data:
                yield item

    async def to_list(self, limit: Optional[int] = None) -> List[T]:
        """Collects items across pages, up to ``limit`` if given."""
        return await _collect(self.__aiter__(), limit)


class AsyncPagePromise(Generic[T]):
    """
    What async ``list()`` returns: ``await`` it for one page, or ``async for``
    over it for every item.
    """

    def __init__(self, load: Callable[[], Awaitable[Any]]) -> None:
        self._load = load

    def __await__(self) -> Generator[Any, None, AsyncPage[T]]:
        result: AsyncPage[T] = yield from self._load().__await__()
        return result

    async def __aiter__(self) -> AsyncIterator[T]:
        page: AsyncPage[T] = await self
        async for item in page:
            yield item

    async def to_list(self, limit: Optional[int] = None) -> List[T]:
        """Collects items across pages, up to ``limit`` if given."""
        return await _collect(self.__aiter__(), limit)


async def _collect(source: AsyncIterator[T], limit: Optional[int]) -> List[T]:
    out: List[T] = []
    if limit is not None and limit <= 0:
        return out
    async for item in source:
        out.append(item)
        if limit is not None and len(out) >= limit:
            break
    return out
